packages/type-utils/tests/isTypeReadonly.test.ts

```diff
@@ -0,0 +1,259 @@
+          const { type: receiver, checker } = getType(
+            `type Test = ReadonlyMap<string, string>;`,
+          );
+
+          const result = isTypeReadonly(checker, receiver);
+          expect(result).toBe(false);
+        });
+      });
+    });
+
+    describe('Intersection', () => {
+      describe('is readonly', () => {
+        it('handles an intersection of 2', () => {
+          const { type, checker } = getType(
+            `type Test = Readonly<{ foo: string; bar: number; }> & Readonly<{ bar: number; }>;`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(true);
+        });
+      });
+
+      describe('is not readonly', () => {
+        it('fails with an intersection of mutables', () => {
+          const { type, checker } = getType(
+            `type Test = { foo: string; bar: number; } & { bar: number; };`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+
+        it('fails with an intersection of mutable to readonly', () => {
+          const { type, checker } = getType(
+            `type Test = { foo: string; bar: number; } & Readonly<{ bar: number; }>;`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+
+        it('fails with an intersection of readonly to mutable', () => {
+          const { type, checker } = getType(
+            `type Test = { foo: string; bar: number; } & Readonly<{ bar: number; }>;`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+      });
+    });
+
+    describe('Union', () => {
+      describe('is readonly', () => {
+        it('handles a union of 2', () => {
+          const { type, checker } = getType(
+            `type Test = readonly string[] | readonly number[];`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(true);
+        });
+      });
+
+      describe('is not readonly', () => {
+        it('fails with a union of mutables', () => {
+          const { type, checker } = getType(`type Test = string[] | number[];`);
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+
+        it('fails with a union of mutable to readonly', () => {
+          const { type, checker } = getType(
+            `type Test = string[] | readonly number[];`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+
+        it('fails with a union of readonly to mutable', () => {
+          const { type, checker } = getType(
+            `type Test = readonly string[] | number[];`,
+          );
+
+          const result = isTypeReadonly(checker, type);
+          expect(result).toBe(false);
+        });
+      });
+    });
+  });
```

Comment:
These tests are super exhaustive and I love them, thank you!

There's a lot of duplication that makes it hard to scan through & see what's under test though. What do you think about [`it.each`](https://jestjs.io/docs/api#testeachtablename-fn-timeout) to condense it a bunch?

```ts
test.each([
  [`{ readonly bar: string; }`, true],
  [`Readonly<{ bar: string; }>;`, true],
  // ...
]);
```

Reply:
I reworked #4416 to make to use this style.